Feature/show qos incompatibility

trittsv opened the pull request:
This Pull-Request adds the possibility to display basic qos incompatibilities between readers and writers on a topic.

The gui looks like this:
<img width="1412" alt="Screenshot 2024-05-26 at 17 51 55" src="https://github.com/eclipse-cyclonedds/cyclonedds-insight/assets/45872415/c3acddbd-f0e7-489e-8988-75297c0674b5">

The yellow warning triangles besides the domain and topic indicates the user that there is a mismatch.
Besides that in the topic view readers and writes with yellow rectangles show which are affected.
When the user hovers over a reader or writer then the specific incompatibility is shown in the tool-tip text.

What it currently cant do is qos xtypes check, this will be implemented in a separate pull request.
To try to keep the pull requests relatively small.

@eboasson can you have a look into this pull request?

eboasson commented on src/dds_service.py:
```diff
@@ -72,3 +74,315 @@ def builtin_observer(domain_id, dds_data, running):
+
+    reliability_rd = to_kind_reliability(endpoint_reader.qos)
+    reliability_wr = to_kind_reliability(endpoint_writer.qos)
+    if reliability_rd and reliability_wr and reliability_rd > reliability_wr:
```
I think it would be nicer if this returned a list of mismatching QoS (perhaps it would then no longer be called `qos_match`, but that's a detail): usually there'll only be a single QoS causing the mismatch but it could potentially be multiple.

I also wonder if it wouldn't be more practical to return the actual mismatching QoS objects instead of a policy id. We have these sparse QoS objects, so why not return a pair of QoS objects with just the mismatching ones filled in? Then the set of policies in those two objects would always be the same and converting to text would be a simple iteration over one and a conversion to text for both.

There's quite a decent chance that I overlooked something. Perhaps a show stopper, or perhaps it turns out to be more complex in practice.

trittsv replied:
thats true, currently it just returns a single mismatch reason, i will change it so that all mismatches can be seen at once.

trittsv replied:
mhm it would not work for all cases, its not always a qos object, sometimes its e.g. topic_name or topic_type..
I think i keep returning the policy id in a list.
In the gui i already have the qos object which can i show when its marked wit a mismatching policy id.

eboasson commented on src/dds_service.py:
```diff
@@ -72,3 +74,315 @@ def builtin_observer(domain_id, dds_data, running):
     logging.info(f"builtin_observer({domain_id}) ... DONE")
+
+
+class dds_durability_kind(OrderedEnum):
```
I wonder if it would be possible to define the ordering on the `Policy` classes directly. I really don't know Python well enough to know whether that's a good idea ... but it seems to be able to compare two, e.g., `Durability` policies or `Liveliness` policies.

(It it makes sense to do that, the ordering should probably not be defined for non-RxO QoS.)

trittsv replied:
i think it should be possible to add the compare operator to the Policy classes.
Lets keep it here for now until all the other qos code from here found its way into the python wrapper?

eboasson replied:
See above, I'm fine with going like this for now.

eboasson commented on src/dds_data.py:
```diff
@@ -103,15 +106,45 @@ def remove_domain_participant(self, domain_id: int, participant: DcpsParticipant
+                            endpoint.key != endpoint_iter.key and
+                            entity_type_iter != entity_type):
+
+                        if entity_type.READER and entity_type_iter.WRITER:
```
Wouldn't it be practical to define class structure for where one has `Endpoint` with two subclasses: `Reader` and `Writer`? Then you can use `isinstance` and the like instead of having to carry the `entity_type` around. (The IDL/XTypes/type stuff in the Python binding does that and it seems to work quite nicely.)

I only see this detail now. I think making such a change would affect quite a bit more code. Even if this is a good idea, it means it doesn't make sense to change it as part of add the QoS compatibility checking.

trittsv replied:
thats a good idea but i will do it in the next pull-request. Because i want to change something in the data handling there anyway - that would fit good together with the changes i want to make there next.

eboasson commented on src/dds_service.py:
```diff
@@ -72,3 +74,315 @@ def builtin_observer(domain_id, dds_data, running):
+
+
+def to_kind_reliability(q: qos.Qos):
+    if qos.Policy.Reliability in q:
```
Python really should have a shorthand notation for this pattern ... but I guess it is write once, check once, run forever ...

trittsv replied:
yes i guess, i did not found a more pretty solution for that, i am open for ides. It will never change so should be ok

eboasson commented on src/dds_service.py:
```diff
@@ -72,3 +74,315 @@ def builtin_observer(domain_id, dds_data, running):
+    DDS_TYPE_CONSISTENCY_DISALLOW_TYPE_COERCION = 0
+    DDS_TYPE_CONSISTENCY_ALLOW_TYPE_COERCION = 1
+
+class dds_qos_policy_id(Enum):
```
These `POLICY_ID` definitions seem out-of-place here. It would make more sense for `qos.py` in the python binding to define them as part of the class definitions. Similarly, it would make more sense for the `OrderedEnum`s defined above to be used by `qos.py` to do the conversion between Python and C. (It currently uses the numbers, not symbolic constants.)

The advantage of defining everything here is of course not having to change the binding for the tool, but in this case, I think it simply shows that the binding took a shortcut too many.

eboasson replied:
Let's keep it for now, add it to the Python binding later. Keeping it here for now means you have more freedom and can move faster 😀

eboasson reviewed (commented):
Nice! I have a few comments relating to some implementation choices and whether it makes sense to do all this in the tool or to push some of the changes in to the Python binding. Perhaps you can give these comments some thought?

trittsv commented on src/dds_data.py:
```diff
@@ -154,6 +189,10 @@ def remove_endpoint(self, domain_id: int, endpoint: DcpsEndpoint):
                     self.remove_topic_signal.emit(domain_id, topic_name)
 
+            # check qos
+            self.no_more_mismatch_in_topic_signal.emit(domain_id, topic_name)
```
this signal is only needed when a endpoint is removed and should not be emitted when a new endpoint arrives. Thats the reason why i did not put it into the `check_qos_mismatches`.

trittsv commented:
@eboasson thank you for your review. I did some modifications - may you can check again?
- i moved the qos code from `dds_service.py` to `dds_qos.py` - which then later can find its way into the python-wrapper
- i changed the `check_qos` from only return one mismatch to return a list with all mismatches (the qos objects are already present in the gui, so only the policy id is needed).
- the `entity_type` will be changed to Endpoint with two subclasses in a other PR (with some other changes i want to make in the data handling)

eboasson reviewed (approved):
LGTM!
